cmd/lakectl/cmd/fs.go

```diff
@@ -409,6 +413,159 @@ var fsRmCmd = &cobra.Command{
 	},
 }
 
+const (
+	fsDownloadCmdMinArgs = 1
+	fsDownloadCmdMaxArgs = 2
+
+	fsDownloadParallelDefault = 6
+)
+
+var fsDownloadCmd = &cobra.Command{
+	Use:   "download <path uri> [<destination path>]",
+	Short: "Download object(s) from a given repository path",
+	Args:  cobra.RangeArgs(fsDownloadCmdMinArgs, fsDownloadCmdMaxArgs),
+	Run: func(cmd *cobra.Command, args []string) {
+		pathURI := MustParsePathURI("path", args[0])
+		flagSet := cmd.Flags()
+		direct := MustBool(flagSet.GetBool("direct"))
+		recursive := MustBool(flagSet.GetBool("recursive"))
+		parallel := MustInt(flagSet.GetInt("parallel"))
+
+		if parallel < 1 {
+			DieFmt("Invalid value for parallel (%d), minimum is 1.\n", parallel)
+		}
+
+		// optional destination directory
+		var dest string
+		if len(args) > 1 {
+			dest = args[1]
+			// verify destination path exists
+			if len(dest) > 0 {
+				pathExists(dest)
+			}
+		}
+
+		// list the files
+		client := getClient()
+		downloadCh := make(chan string)
+		sourcePath := api.StringValue(pathURI.Path)
+
+		// recursive assume source is directory
+		if recursive && len(sourcePath) > 0 && !strings.HasSuffix(sourcePath, "/") {
+			sourcePath += "/"
+		}
+
+		ctx := cmd.Context()
+		// list objects to download
+		go func() {
+			defer close(downloadCh)
+			if recursive {
+				listRecursiveHelper(ctx, client, pathURI.Repository, pathURI.Ref, sourcePath, downloadCh)
+			} else {
+				downloadCh <- api.StringValue(pathURI.Path)
+			}
+		}()
+
+		// download in parallel
+		var (
+			wg         sync.WaitGroup
+			errCounter int64
+		)
+		wg.Add(parallel)
+		for i := 0; i < parallel; i++ {
+			go func() {
+				defer wg.Done()
+				for downloadPath := range downloadCh {
+					src := uri.URI{
+						Repository: pathURI.Repository,
+						Ref:        pathURI.Ref,
+						Path:       &downloadPath,
+					}
+					// destination is without the source URI
+					dst := filepath.Join(dest, strings.TrimPrefix(downloadPath, sourcePath))
+					err := downloadHelper(ctx, client, direct, src, dst)
+					if err == nil {
+						fmt.Printf("Download: %s to %s\n", src.String(), dst)
```

Comment:
Maybe we shouldn't print on every file - since this might spam the stdout?

Reply:
Preferably print a single 'download from path: ' on recursive flag
Suggest also to add a periodic output the prints:
'Downloaded x/y objects'

Reply:
In this implementation is cloned the `aws s3 cp` behaviour - but I'll check how to use one of the progress packages we have to get line per worker that will print out the current uploaded file.

Reply:
@N-o-Z played with it a little - think we should align the progress between upload/download. The big issue for me was the fact that it was hard to test the output when periodic download progress was printed.
I liked the fact that we are like aws cp and you can log and grep the output.